### PR TITLE
FEAT-#6284: Do not convert HDK query execution result to arrow.

### LINKED ISSUE #6284: FEAT: Do not convert HDK query execution result to arrow.
Currently, HdkExecutor converts the execution results to arrow and returns the arrow table. In case of subsequent query, the table is imported to HDK again. The latest version of HDK stores the result in a temporary table and we can reuse this table for the subsequent queries. It would allow to avoid the unnecessary to/from arrow export/import.

### PR DESCRIPTION
**Pull request by AndreyPavlenko**
## What do these changes do?

Now, the execution result is not converted to arrow but is stored in partitions as the ExecutionResultTable object. This object represents a table in the HDK database and provides some basic information about the table - name, column_names, shape. If an arrow table is imported to HDK, a corresponding ImportedTable object is stored in partitions. All the subsequent operations are performed, using the *Table objects, without conversion to/from arrow.

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6284 
- [] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py [hunk 1]
```diff
@@ -419,15 +428,60 @@ def take_2d_labels_or_positional(
...
-                columns=base.columns,
-                dtypes=base.copy_dtypes_cache(),
-                op=op,
-                index_cols=base._index_cols,
-                force_execution_mode=base._force_execution_mode,
+        if row_positions is None:
+            return base
+
+        # Check if row_positions could be converted to a range.
+        if len(row_positions) > 2 and not is_range_like(row_positions):
```
Comment (Garra1980) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:435]: Is this related to the PR's goal?

Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:435]: It's not. This PR changed execution of some operations, that were previously executed via arrow and now - via HDK. It caused the `asv_bench` failures. The failures occured due to a long range-like list of `row_positions`, that was converted to a huge json query, that HDK failed to execute.

**Comment by Garra1980** [comment]
I think we also need to reflect changes here - https://modin.readthedocs.io/en/stable/flow/modin/experimental/core/execution/native/implementations/hdk_on_native/index.html

**Comment by AndreyPavlenko** [comment]
> I think we also need to reflect changes here - https://modin.readthedocs.io/en/stable/flow/modin/experimental/core/execution/native/implementations/hdk_on_native/index.html

Done.

File: docs/flow/modin/experimental/core/execution/native/implementations/hdk_on_native/index.rst [hunk 2]
```diff
@@ -173,15 +177,15 @@ Arrow execution
 
 For simple operations which don't include actual computations, execution can use
 Arrow API. We can use it to rename columns, drop columns and concatenate
-frames. Arrow execution is preferable since it doesn't require actual data import/export
-from/to HDK.
+frames. Arrow execution is performed if we have an arrow table in the partition
+abd it's preferable since it doesn't require actual data import into HDK.
```
Comment (Garra1980) [inline docs/flow/modin/experimental/core/execution/native/implementations/hdk_on_native/index.rst:181]: misprint

**Comment by Garra1980** [comment]
> > I think we also need to reflect changes here - https://modin.readthedocs.io/en/stable/flow/modin/experimental/core/execution/native/implementations/hdk_on_native/index.html
> 
> Done.

Should we also change the last diagram on this page? Now it's all about ArrowTable at the end of flow

**Comment by AndreyPavlenko** [comment]
> > > I think we also need to reflect changes here - https://modin.readthedocs.io/en/stable/flow/modin/experimental/core/execution/native/implementations/hdk_on_native/index.html
> > 
> > 
> > Done.
> 
> Should we also change the last diagram on this page? Now it's all about ArrowTable at the end of flow

Done.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py [hunk 3]
```diff
@@ -419,15 +428,60 @@ def take_2d_labels_or_positional(
...
+        op = MaskNode(base, row_labels=row_labels, row_positions=row_positions)
+        base = self.__constructor__(
+            columns=base.columns,
+            dtypes=base.copy_dtypes_cache(),
+            op=op,
+            index_cols=base._index_cols,
+            force_execution_mode=base._force_execution_mode,
+        )
+
+        # Reverse the frame rows, if performing a reverse order selection via HDK.
```
Comment (Garra1980) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:455]: Do I and correctly that we sacrifice performance to get this particular case work?

File: modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py [hunk 4]
```diff
@@ -1955,21 +2031,24 @@ def _materialize(self):
...
+        assert mode != "lazy", "Unexpected execution triggered on lazy frame!"
+
+        if isinstance(self._op, FrameNode):
+            return self._op.execute_arrow()
 
-        if self._force_execution_mode != "hdk" and self._can_execute_arrow():
+        if (
+            mode == "arrow"
+            or not self._op.can_execute_hdk()
+            or (self._can_execute_arrow() and mode != "hdk")
```
Comment (Garra1980) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:2045]: I see that _force_execution_mode description only mentions "lazy" and "arrow", but there is also "hdk", right?

File: modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py (hunk 3 above)
Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:455]: Yes, but it seems the only way to reverse rows via HDK.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py (hunk 4 above)
Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:2045]: Added "hdk" to the description.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py (hunk 3 above)
Comment (Garra1980) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:455]: Got it, thanks. ANyway it looks like a subject for submitting a bug in hdk. This can be done later of course

File: modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py [hunk 5]
```diff
@@ -1107,6 +1108,9 @@ def _fold_invert(self) -> Union["OpExpr", LiteralExpr]:
                 return OpExpr("IS NOT NULL", op.operands, op._dtype)
             if op.op == "IS NOT NULL":
                 return OpExpr("IS NULL", op.operands, op._dtype)
+            if op.op == "CASE":
```
Comment (Garra1980) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py:1111]: Again, not related to arrow exporting changes?

Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py:1111]: Yes, it was a test failure fix. I don't remember which one.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py (hunk 3 above)
Comment (ienkovich) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:455]: Does it mean we also completely lose an option to arrow-execute this case?

Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:455]: No, below is the condition - `and not base._can_execute_arrow()`. If we have an arrow table in the partition, it will be executed with arrow.

**Comment by anmyachev** [comment]
@AndreyPavlenko CodeQL failed

**Comment by AndreyPavlenko** [comment]
> @AndreyPavlenko CodeQL failed

I saw the failures, but I don't see the import cycle there. 
The 'variable may be used before it is initialized' errors seem false positive.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py [hunk 6]
```diff
@@ -12,94 +12,157 @@
...
+
+    def __init__(self, result: ExecutionResult):
+        self.name = result.table_name
+        self._result = result
+
+    @property
+    @_inherit_docstrings(DbTable.shape)
+    def shape(self) -> Tuple[int, int]:
+        if (shape := getattr(self, "_shape", None)) is None:
+            self._shape = shape = self.scan().shape
```
Comment (ienkovich) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py:45]: You can use `shape` property of `self._result`. A scan is not required.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py [hunk 7]
```diff
@@ -12,94 +12,157 @@
...
+        return shape
+
+    @property
+    @_inherit_docstrings(DbTable.column_names)
+    def column_names(self) -> List[str]:
+        if (names := getattr(self, "_column_names", None)) is None:
+            self._column_names = names = list(self.scan().schema)
```
Comment (ienkovich) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py:52]: `schema` is also available for `self._result`

File: docs/conf.py [hunk 8]
```diff
@@ -24,7 +24,7 @@ def noop_decorator(*args, **kwargs):
 ray.remote = noop_decorator
 
 # fake modules if they're missing
-for mod_name in ("cudf", "cupy", "pyarrow.gandiva", "pyhdk"):
+for mod_name in ("cudf", "cupy", "pyarrow.gandiva", "pyhdk", "pyhdk.hdk", "pyhdk._sql"):
```
[the remark this answers is not in the document]
Comment (AndreyPavlenko) [reply docs/conf.py:27]: The symbols are available, but only if pyhdk is installed. As I understand, these fake modules are used in the modin_dev environment, that does not have pyhdk.

[the remark this answers is not in the document]
Comment (ienkovich) [reply docs/conf.py:27]: If you were importing these classes from the top-level `pyhdk` then you would need to add these fake classes to `pyhdk` module, not to its sub-modules. `pyhdk._sql` is a hidden submodule and is not supposed to be used directly.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py [hunk 9]
```diff
@@ -1107,6 +1108,9 @@ def _fold_invert(self) -> Union["OpExpr", LiteralExpr]:
...
+                operands = [op.operands[0], op.operands[2], op.operands[1]]
```
[the remark this answers is not in the document]
Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py]: I thought the condition should be inverted, but I was wrong. Thanks for pointing it out!

File: modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py [hunk 10]
```diff
@@ -12,94 +12,148 @@
...
 
-    _config = None
-    _storage = None
-    _data_mgr = None
-    _calcite = None
-    _executor = None
+    def __new__(cls, *args, **kwargs):
+        if (instance := getattr(cls, "_instance", None)) is None:
+            cls._instance = instance = object.__new__(cls)
+        return instance
```
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py:116]: Please rewrite this place without using `:=` operator. This will fix this check.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py [hunk 11]
```diff
@@ -12,94 +12,158 @@
...
+    def __init__(self, result: ExecutionResult):
+        self.name = result.table_name
+        self._result = result
+
+    @property
+    @_inherit_docstrings(DbTable.shape)
+    def shape(self) -> Tuple[int, int]:
+        if (shape := getattr(self, "_shape", None)) is None:
+            self._shape = shape = self.scan().shape
+        return shape
```
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py:45]: Please rewrite this place without using `:=` operator. This will fix this check.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py [hunk 12]
```diff
@@ -12,94 +12,158 @@
...
+
+    @property
+    @_inherit_docstrings(DbTable.column_names)
+    def column_names(self) -> List[str]:
+        if (names := getattr(self, "_column_names", None)) is None:
+            self._column_names = names = list(self.scan().schema)
+        return names
```
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py:52]: Please rewrite this place without using `:=` operator. This will fix this check.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py [hunk 13]
```diff
@@ -12,94 +12,158 @@
...
+        QueryNode
         """
-        if cls._executor is None:
-            cls._config = pyhdk.buildConfig(**HdkLaunchParameters.get())
-            cls._storage = pyhdk.storage.ArrowStorage(1)
-            cls._data_mgr = pyhdk.storage.DataMgr(cls._config)
-            cls._data_mgr.registerDataProvider(cls._storage)
+        if (scan := getattr(self, "_scan", None)) is None:
+            self._scan = scan = HdkWorker._hdk().scan(self._result.table_name)
+        return scan
```
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py:73]: Please rewrite this place without using `:=` operator. This will fix this check.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py:116]: Could we configure CodeQL somehow? Limiting the language capabilities just to fix a false positive error seems not reasonable to me.

[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py:116]: Possibly, but we can definitely dismiss these alerts. However, in this case, `instance` variable is used outside of the `if` scope. In this case, the code is easier to read with a regular assignment operator.

```python
        instance = getattr(cls, "_instance", None)
        if instance is None:
            cls._instance = instance = object.__new__(cls)
        return instance
```

File: modin/experimental/core/execution/native/implementations/hdk_on_native/interchange/dataframe_protocol/dataframe.py [hunk 14]
```diff
@@ -209,7 +209,8 @@ def _pyarrow_table(self) -> pa.Table:
         -------
         pyarrow.Table
         """
-        at = self._df._execute()
+        if not isinstance(at := self._df._execute(), pa.Table):
+            at = at.to_arrow()
```
Comment (anmyachev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/interchange/dataframe_protocol/dataframe.py:213]:
```suggestion
        at = self._df._execute()
        if not isinstance(at, pa.Table):
            at = at.to_arrow()
```

File: modin/experimental/sql/hdk/query.py [hunk 15]
```diff
@@ -10,7 +10,6 @@
 # the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
 # ANY KIND, either express or implied. See the License for the specific language
 # governing permissions and limitations under the License.
-
```
Comment (anmyachev) [inline modin/experimental/sql/hdk/query.py:13]: pls revert

File: modin/test/interchange/dataframe_protocol/hdk/test_protocol.py [hunk 16]
```diff
@@ -47,6 +48,17 @@ def test_simple_export(data_has_nulls, from_hdk, n_chunks):
     md_df = pd.DataFrame(data)
 
     exported_df = export_frame(md_df, from_hdk, n_chunks=n_chunks)
+
+    if StorageFormat.get() == "Hdk":
```
Comment (anmyachev) [inline modin/test/interchange/dataframe_protocol/hdk/test_protocol.py:52]: How did it work before?

Why check the `StorageFormat` at all if these tests are specific to hdk?

File: modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py [hunk 17]
```diff
@@ -463,8 +517,11 @@ def _dtypes_for_exprs(self, exprs):
     def _maybe_update_proxies(self, dtypes, new_parent=None):
         if new_parent is not None:
             super()._maybe_update_proxies(dtypes, new_parent)
-        elif self._has_arrow_table():
-            table = self._partitions[0, 0].get()
+        if self._partitions is None:
+            return
+        if isinstance(table := self._partitions[0][0].get(), ImportedTable):
+            table = table.to_arrow()
+        if isinstance(table, pyarrow.Table):
```
Comment (anmyachev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:524]:
```suggestion
        table = self._partitions[0][0].get()
        if isinstance(table, ImportedTable):
            table = table.to_arrow()
        if isinstance(table, pyarrow.Table):
```

File: modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py (hunk 1 above)
Comment (anmyachev) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:435]: @AndreyPavlenko It's better to put it in a separate function.

**Comment by anmyachev** [comment]
@AndreyPavlenko need to resolve conflicts

File: modin/test/interchange/dataframe_protocol/hdk/test_protocol.py (hunk 16 above)
Comment (AndreyPavlenko) [reply modin/test/interchange/dataframe_protocol/hdk/test_protocol.py:52]: Before it was executed with arrow.

The check is not required here, thanks.

Comment (anmyachev) [reply modin/test/interchange/dataframe_protocol/hdk/test_protocol.py]: Is there any chance to quickly fix this by defaulting to arrow?

**Review by anmyachev — APPROVED** [review]
LGTM!

File: modin/test/interchange/dataframe_protocol/hdk/test_protocol.py (hunk 16 above)
Comment (AndreyPavlenko) [reply modin/test/interchange/dataframe_protocol/hdk/test_protocol.py]: No. The main idea of this PR is to minimize the imports/exports from/to arrow. If a table has been imported into HDK, the subsequent execution should be performed on the HDK side.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/partitioning/partition.py [hunk 18]
```diff
@@ -124,3 +107,28 @@ def put(cls, obj):
...
+        Number of columns.
+
+        Returns
+        -------
+        int
+        """
+        if isinstance(self._data, pa.Table):
+            return self._data.num_columns
+        else:
+            return self._data.shape[0]
```
Comment (YarShev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/partitioning/partition.py:134]: is shape[0] the correct way to access the number of columns? not shape[1]?

Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/partitioning/partition.py:134]: You are right, it should be shape[1]. I'll fix it in a separate PR.

Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/partitioning/partition.py:134]: https://github.com/modin-project/modin/pull/6353
